network/src/network_config.rs

```diff
@@ -1,93 +1,102 @@
 use crate::errors::{ConfigError, Error};
 use crate::PeerId;
-use bytes::Bytes;
 use log::info;
 use p2p::multiaddr::{Multiaddr, Protocol, ToMultiaddr};
 use rand;
 use rand::Rng;
 use secio;
+use serde_derive::Deserialize;
 use std::fs;
-use std::io::Error as IoError;
 use std::io::Read;
 use std::io::Write;
-use std::iter;
-use std::net::Ipv4Addr;
-use std::time::Duration;
+use std::path::PathBuf;
 
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, Deserialize)]
 pub struct NetworkConfig {
     pub listen_addresses: Vec<Multiaddr>,
     pub public_addresses: Vec<Multiaddr>,
-    pub client_version: String,
-    pub protocol_version: String,
+    pub bootnodes: Vec<Multiaddr>,
+    pub reserved_peers: Vec<Multiaddr>,
     pub reserved_only: bool,
-    pub max_inbound_peers: u32,
+    pub max_peers: u32,
     pub max_outbound_peers: u32,
-    pub reserved_peers: Vec<String>,
-    pub secret_key: Option<Bytes>,
-    pub secret_key_path: Option<String>,
-    pub peer_store_path: Option<String>,
-    pub config_dir_path: Option<String>,
-    pub bootnodes: Vec<String>,
-    pub ping_interval: Duration,
-    pub ping_timeout: Duration,
-    pub discovery_timeout: Duration,
-    pub discovery_response_count: usize,
-    pub discovery_interval: Duration,
-    pub try_outbound_connect_interval: Duration,
+    pub path: PathBuf,
+    pub ping_interval_secs: u64,
+    pub ping_timeout_secs: u64,
+    pub connect_outbound_interval_secs: u64,
 }
 
-impl NetworkConfig {
-    pub(crate) fn read_secret_key(&self) -> Option<Bytes> {
-        if self.secret_key.is_some() {
-            self.secret_key.clone()
-        } else if let Some(ref path) = self.secret_key_path {
-            match fs::File::open(path).and_then(|mut file| {
-                let mut buf = Vec::new();
-                file.read_to_end(&mut buf).map(|_| buf)
-            }) {
-                Ok(secret) => Some(secret.into()),
-                Err(_err) => None,
-            }
-        } else {
-            None
+fn generate_random_key() -> [u8; 32] {
+    loop {
+        let mut key: [u8; 32] = [0; 32];
+        rand::thread_rng().fill(&mut key);
+        if secio::SecioKeyPair::secp256k1_raw_key(&key).is_ok() {
+            return key;
         }
     }
+}
 
-    pub fn new() -> Self {
-        Self::default()
+impl NetworkConfig {
+    pub fn secret_key_path(&self) -> PathBuf {
+        let mut path = self.path.clone();
+        path.set_file_name("secret_key");
```

Comment:
I think that we should use `push` here, `set_file_name` will replace the folder, please check:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=0bb6b29be629e34476bce4d091c6669d

Reply:
fixed

Reply:
fixed